src/python/pants/backend/python/target_types_rules.py

```diff
@@ -433,6 +441,70 @@ async def inject_python_distribution_dependencies(
     )
 
 
+# -----------------------------------------------------------------------------------------------
+# Dependency validation
+# -----------------------------------------------------------------------------------------------
+
+
+@dataclass(frozen=True)
+class DependencyValidationFieldSet(FieldSet):
+    required_fields = (InterpreterConstraintsField,)
+
+    interpreter_constraints: InterpreterConstraintsField
+
+
+class PythonValidateDependenciesRequest(ValidateDependenciesRequest):
+    field_set_type = DependencyValidationFieldSet
+
+
+@rule
+async def validate_python_dependencies(
+    request: PythonValidateDependenciesRequest,
+    python_setup: PythonSetup,
+) -> ValidatedDependencies:
+    dependencies = await MultiGet(Get(WrappedTarget, Address, d) for d in request.dependencies)
+
+    # Validate that the ICs for dependencies are all compatible with our own.
+    target_ics = request.field_set.interpreter_constraints.value_or_global_default(python_setup)
+    non_subset_items = []
+    for dep in dependencies:
+        if not dep.target.has_field(InterpreterConstraintsField):
+            continue
+        dep_ics = dep.target[InterpreterConstraintsField].value_or_global_default(python_setup)
+        if not interpreter_constraints_contains(
+            dep_ics, target_ics, python_setup.interpreter_universe
+        ):
+            non_subset_items.append(f"{dep_ics}: {dep.target.address}")
+
+    # TODO: When this deprecation triggers, it should be converted into an exception, and
+    # all usages of the InterpreterConstraints methods:
+    #   * compute_for_targets
+    #   * create_from_compatibility_fields
+    #   * create_from_targets
+    # ... should be replaced with calls to InterpreterConstraintsField.value_or_global_default.
+    deprecated_conditional(
+        lambda: bool(non_subset_items),
+        removal_version="2.14.0.dev1",
```

Comment:
Since this gets cherry-picked to 2.12, how about removing in 2.13? We can then make good performance enhancements etc

Ack that cherry-picks won't be as fun. Imo that's worth it.